src/Training.AirBnb.Clone.Backend/Backend Project/Infrastructure/Services/ListingServices/ListingOccupancyService.cs

```diff
@@ -0,0 +1,79 @@
+    private readonly IDataContext _appDateContext;
+
+    public ListingOccupancyService(IDataContext appDataContext)
+    {
+        _appDateContext = appDataContext;        
+    }
+
+    public async ValueTask<ListingOccupancy> CreateAsync(ListingOccupancy occupancy, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
+        if (!IsValidOccupancy(occupancy))
+            throw new EntityValidationException<ListingOccupancy>();
+
+        await _appDateContext.ListingOccupancies.AddAsync(occupancy, cancellationToken);
+
+        if (saveChanges) await _appDateContext.ListingOccupancies.SaveChangesAsync(cancellationToken);
+
+        return occupancy;
+    }
+
+    public ValueTask<ICollection<ListingOccupancy>> GetAsync(IEnumerable<Guid> ids, CancellationToken cancellationToken = default)
+        => new ValueTask<ICollection<ListingOccupancy>>(GetUndeletedListingOccupancies()
+            .Where(occupancy => ids.Contains(occupancy.Id))
+            .ToList());
+
+    public ValueTask<ListingOccupancy> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
+        => new ValueTask<ListingOccupancy>(GetUndeletedListingOccupancies()
+            .FirstOrDefault(occupancy => occupancy.Id == id)
+            ?? throw new EntityNotFoundException<ListingOccupancy>());
+
+    public IQueryable<ListingOccupancy> Get(Expression<Func<ListingOccupancy, bool>> predicate)
+        => GetUndeletedListingOccupancies().Where(predicate.Compile()).AsQueryable();
+
+    public async ValueTask<ListingOccupancy> UpdateAsync(ListingOccupancy occupancy, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
+        if (!IsValidOccupancy(occupancy))
+            throw new EntityValidationException<ListingOccupancy>();
+
+        var foundOccupancy = await GetByIdAsync(occupancy.Id, cancellationToken);
+
+        foundOccupancy.Guests = occupancy.Guests;
+        foundOccupancy.AllowPets = occupancy.AllowPets;
+        foundOccupancy.ModifiedDate = DateTime.UtcNow;
+
+        if (saveChanges) await _appDateContext.ListingOccupancies.SaveChangesAsync(cancellationToken);
+
+        return foundOccupancy;
+    }
+
+    public async ValueTask<ListingOccupancy> DeleteAsync(Guid id, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
+        var foundOccupancy = await GetByIdAsync(id, cancellationToken);
+
+        foundOccupancy.IsDeleted = true;
+        foundOccupancy.DeletedDate = DateTime.UtcNow;
+
+        if (saveChanges) await _appDateContext.ListingOccupancies.SaveChangesAsync(cancellationToken);
+
+        return foundOccupancy;
+    }
+
+    public async ValueTask<ListingOccupancy> DeleteAsync(ListingOccupancy occupancy, bool saveChanges = true, CancellationToken cancellationToken = default)
+        => await DeleteAsync(occupancy.Id, saveChanges, cancellationToken);
+
+    private bool IsValidOccupancy(ListingOccupancy occupancy)
+        => occupancy.Guests >= 1;
+
```

Comment:
maxsimum chegarani tekshirish kerak emasmi shu yerda

Reply:
Maksimum chegara har bir listing categoriyasida har xil bo'ladi. Shuning uchun maksimu chegara crud service da emas composition service da tekshiriladi.

Reply:
Bo'lmasam crud service ga boshqa bir crud service ni inject qilishga to'g'ri keladi. Undan ko'ra composition service qilgan afzal.